rmw_zenoh_cpp/src/rmw_client.cpp

```diff
@@ -0,0 +1,465 @@
+  if (strlen(service_name) == 0) {
+    RMW_SET_ERROR_MSG("service name is empty string");
+    return nullptr;
+  }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(qos_profile, nullptr);
+
+  // TODO(CH3): When we figure out how to spoof QoS, check for a 'configured' QoS to pass the final
+  // test that is failing
+  //
+  // if (# SOME CHECK HERE) {
+  //   RMW_SET_ERROR_MSG("expected configured QoS profile");
+  //   return nullptr;
+  // }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(type_supports, nullptr);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  // VALIDATE SERVICE NAME =====================================================
+  int * validation_result = static_cast<int *>(allocator->allocate(sizeof(int), allocator->state));
+  rmw_validate_full_topic_name(service_name, validation_result, nullptr);
+
+  if (*validation_result == RMW_TOPIC_VALID || qos_profile->avoid_ros_namespace_conventions) {
+    allocator->deallocate(validation_result, allocator->state);
+  } else {
+    RMW_SET_ERROR_MSG("publisher topic is malformed!");
+    allocator->deallocate(validation_result, allocator->state);
+    return nullptr;
+  }
+
+  // OBTAIN TYPESUPPORT ========================================================
+  const rosidl_service_type_support_t * type_support = get_service_typesupport_handle(
+    type_supports, RMW_ZENOH_CPP_TYPESUPPORT_C
+  );
+
+  if (!type_support) {
+    type_support = get_service_typesupport_handle(
+      type_supports, RMW_ZENOH_CPP_TYPESUPPORT_CPP);
+    if (!type_support) {
+      RCUTILS_LOG_INFO("%s", service_name);
+      RMW_SET_ERROR_MSG("type support not from this implementation");
+      return nullptr;
+    }
+  }
+
+  // CREATE CLIENT =============================================================
+  rmw_client_t * client = static_cast<rmw_client_t *>(allocator->allocate(sizeof(rmw_client_t),
+                                                                          allocator->state));
+  if (!client) {
+    RMW_SET_ERROR_MSG("failed to allocate rmw_client_t");
+    allocator->deallocate(client, allocator->state);
+    return nullptr;
+  }
+
+  // Populate common members
+  client->implementation_identifier = eclipse_zenoh_identifier;  // const char * assignment
+
+  client->service_name = rcutils_strdup(service_name, *allocator);  // const char * assignment
+  if (!client->service_name) {
+    RMW_SET_ERROR_MSG("failed to allocate publisher topic name");
+    allocator->deallocate(client, allocator->state);
+    return nullptr;
+  }
+
+  client->data = static_cast<rmw_client_data_t *>(allocator->allocate(sizeof(rmw_client_data_t),
+                                                                      allocator->state));
+  if (!client->data) {
+    RMW_SET_ERROR_MSG("failed to allocate client data");
+    allocator->deallocate(client->data, allocator->state);
+    allocator->deallocate(client, allocator->state);
```

Comment:
Does this leak `client->service_name`?

Reply:
Yep!

Updated: 23497d4430f890dc62fb9995c66fb2be8602b51e

(Sorry for the massive commit, it's kind of intertwined...)